core/src/main/java/com/linecorp/armeria/client/retry/RetryingClient.java

```diff
@@ -282,16 +284,26 @@ private void doExecute0(ClientRequestContext ctx, HttpRequestDuplicator rootReqD
             duplicateReq = rootReqDuplicator.duplicate(newHeaders.build());
         }
 
-        final ClientRequestContext derivedCtx;
+        final DefaultClientRequestContext derivedCtx;
         try {
-            derivedCtx = newDerivedContext(ctx, duplicateReq, ctx.rpcRequest(), initialAttempt);
+            derivedCtx = newDerivedContext(ctx, duplicateReq, ctx.rpcRequest(), initialAttempt)
+                    .as(DefaultClientRequestContext.class);
         } catch (Throwable t) {
             handleException(ctx, rootReqDuplicator, future, t, initialAttempt);
             return;
         }
-
-        final HttpResponse response = executeWithFallback(unwrap(), derivedCtx,
-                                                          (context, cause) -> HttpResponse.ofFailure(cause));
+        assert derivedCtx != null;
+
+        final HttpResponse response;
+        if (derivedCtx.endpointGroup() != null && derivedCtx.endpoint() == null) {
+            // if the endpoint hasn't been selected, try to initialize the ctx with a new endpoint/event loop
+            response = initContextAndExecuteWithFallback(unwrap(), derivedCtx, derivedCtx.endpointGroup(),
```

Comment:
I've added an assertion for this case which passes 😅 
https://github.com/line/armeria/blob/e0583b3e1eb97e1e0e5cdc61188ce5c79e199885/core/src/test/java/com/linecorp/armeria/client/retry/RetryingClientWithEmptyEndpointGroupTest.java#L162

Currently, I think `DefaultClientRequestContext#customizer` isn't actually copied over to derived contexts (although `options` is copied over). I think this may be a bug - I would like to handle this issue separately with a separate issue if this sounds good to you (since this case is guarded by a unit test).